engine/runtime/src/main/java/org/enso/interpreter/node/expression/constant/LazyObjectNode.java

```diff
@@ -0,0 +1,44 @@
+package org.enso.interpreter.node.expression.constant;
+
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.interop.TruffleObject;
+import com.oracle.truffle.api.nodes.NodeInfo;
+import java.util.function.Supplier;
+import org.enso.interpreter.node.ExpressionNode;
+import org.enso.interpreter.runtime.data.text.Text;
+import org.enso.interpreter.runtime.error.DataflowError;
+import org.enso.interpreter.runtime.util.CachingSupplier;
+
+@NodeInfo(
+    shortName = "lazy",
+    description = "Represents an arbitrary compile-time constant computed lazily.")
+public final class LazyObjectNode extends ExpressionNode {
+
+  private final String error;
+  private final CachingSupplier<? extends Object> supply;
+
+  private LazyObjectNode(String error, Supplier<? extends Object> supply) {
+    this.error = error;
+    this.supply = CachingSupplier.wrap(supply);
+  }
+
+  /**
+   * Creates a node that returns lazily computed value.
+   *
+   * @param errorMessage the error message to show when the value is {@code null}
+   * @param supplier computes the value lazily. Can return {@code null} and then the {@code
+   *     errorMessage} error is created
+   */
+  public static ExpressionNode build(String errorMessage, Supplier<TruffleObject> supplier) {
+    return new LazyObjectNode(errorMessage, supplier);
+  }
+
+  @Override
+  public Object executeGeneric(VirtualFrame frame) {
+    var result = supply.get();
+    if (result instanceof TruffleObject) {
+      return result;
+    }
+    return DataflowError.withDefaultTrace(Text.create(error), this);
```

Comment:
You only check if it is a `TruffleObject` not if it is `null` which seems to be what javadoc suggests?

Reply:
Do you mean the Javadoc of [build method](https://github.com/enso-org/enso/pull/10899/files/53eece8f3a30231a9f095714ec2b3120b517a2ba#diff-7811c280da1db3c47bf9c385c5d9453840df54721b64d672425f1745e7ddc28bR32) that takes `Supplier<TruffleObject>` argument?

If typing is on, then check for `null` is the same as `instanceof TruffleObject`. But I decided to go for the latter as someone could sneak in raw `Supplier` and return anything. I would need to perform some complicated checks to verify the value is proper _Enso interpreter value_. That can be added, but for now `instanceof TruffleObject` is sufficient.